pkg/query-service/agentConf/db.go

```diff
@@ -0,0 +1,224 @@
+package agentConf
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"math/rand"
+
+	"github.com/google/uuid"
+	"github.com/jmoiron/sqlx"
+	"go.signoz.io/signoz/pkg/query-service/agentConf/sqlite"
+	"go.signoz.io/signoz/pkg/query-service/model"
+	"go.uber.org/zap"
+)
+
+func init() {
+	rand.Seed(2000)
+}
+
+// Repo handles DDL and DML ops on ingestion rules
+type Repo struct {
+	db *sqlx.DB
+}
+
+func (r *Repo) initDB(engine string) error {
+	switch engine {
+	case "sqlite3", "sqlite":
+		return sqlite.InitDB(r.db)
+	default:
+		return fmt.Errorf("unsupported db")
+	}
+}
+
+func (r *Repo) GetConfigHistory(ctx context.Context, typ ElementTypeDef) ([]ConfigVersion, error) {
+	var c []ConfigVersion
+	err := r.db.SelectContext(ctx, &c, `SELECT 
+		id, 
+		version, 
+		element_type, 
+		COALESCE(created_by, -1) as created_by, 
+		active, 
+		is_valid, 
+		disabled, 
+		deploy_status, 
+		deploy_result 
+		FROM agent_config_versions 
+		WHERE element_type = $1`, typ)
+
+	return c, err
+}
+
+func (r *Repo) GetConfigVersion(ctx context.Context, typ ElementTypeDef, v int) (*ConfigVersion, error) {
+	var c ConfigVersion
+	err := r.db.GetContext(ctx, &c, `SELECT 
+		id, 
+		version, 
+		element_type, 
+		COALESCE(created_by, -1) as created_by, 
+		COALESCE((SELECT NAME FROM users 
+		WHERE id = v.created_by), "unknown") created_by_name,
+		active, 
+		is_valid, 
+		disabled, 
+		deploy_status, 
+		deploy_result,
+		last_hash,
+		last_config
+		FROM agent_config_versions v 
+		WHERE element_type = $1 
+		AND version = $2`, typ, v)
+
+	return &c, err
+
+}
+
+func (r *Repo) GetLatestVersion(ctx context.Context, typ ElementTypeDef) (*ConfigVersion, error) {
+	var c ConfigVersion
+	err := r.db.GetContext(ctx, &c, `SELECT 
+		id, 
+		version, 
+		element_type, 
+		COALESCE(created_by, -1) as created_by, 
+		active, 
+		is_valid, 
+		disabled, 
+		deploy_status, 
+		deploy_result 
+		FROM agent_config_versions 
+		WHERE element_type = $1 
+		AND version = ( 
+			SELECT MAX(version) 
+			FROM agent_config_versions 
+			WHERE element_type=$2)`, typ, typ)
+	if err != nil {
+		zap.S().Errorf("failed get latest config version for element:", typ, err)
+	}
+	return &c, err
+}
+
+func (r *Repo) insertConfig(ctx context.Context, c *ConfigVersion, elements []string) (fnerr error) {
+
+	if string(c.ElementType) == "" {
+		return fmt.Errorf("element type is required for creating agent config version")
+	}
+
+	if len(elements) == 0 {
+		zap.S().Errorf("insert config called with no elements", c.ElementType)
+		return fmt.Errorf("config must have atleast one element")
+	}
+
+	if c.Version != 0 {
```

Comment:
This is the struct 
ConfigVersion {
  Version int
} 

The versions in insertConfig need to be auto-assigned and monotonically increased. Hence if InsertConfig is called with a set version we reject the request. The version has to be unassigned. 
Could have used *int instead of int but the versions start with 1. So this is still workable.

Reply:
Ok I will add a comment in the code to explain